File: src/main/java/com/demonwav/statcraft/commands/sc/SCDamageTaken.java
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * http://demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc;

import com.demonwav.statcraft.StatCraft;
import com.demonwav.statcraft.commands.ResponseBuilder;
import com.demonwav.statcraft.querydsl.QDamageTaken;
import com.demonwav.statcraft.querydsl.QPlayers;

import com.mysema.query.Tuple;
import com.mysema.query.sql.SQLQuery;
import org.bukkit.command.CommandSender;

import java.sql.Connection;
import java.util.List;

public class SCDamageTaken extends SCTemplate {

    public SCDamageTaken(StatCraft plugin) {
        super(plugin);
        this.plugin.getBaseCommand().registerCommand("damagetaken", this);
    }

    @Override
    public boolean hasPermission(CommandSender sender, String[] args) {
        return sender.hasPermission("statcraft.user.damagetaken");
    }

    @Override
    public String playerStatResponse(String name, List<String> args, Connection connection) {
        try {
            int id = getId(name);
            if (id < 0)
                throw new Exception();

            SQLQuery query = plugin.getDatabaseManager().getNewQuery(connection);
            if (query == null)
                return "Sorry, there seems to be an issue connecting to the database right now.";
            QDamageTaken d = QDamageTaken.damageTaken;
            Integer total = query.from(d).where(d.id.eq(id)).uniqueResult(d.amount.sum());

            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Damage Taken")
                    .addStat("Total", df.format(total))
                    .toString();
        } catch (Exception e) {
            return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Damage Taken")
                    .addStat("Total", String.valueOf(0))
                    .toString();
        }
    }

    @Override
    public String serverStatListResponse(int num, List<String> args, Connection connection) {
        SQLQuery query = plugin.getDatabaseManager().getNewQuery(connection);
        if (query == null)
            return "Sorry, there seems to be an issue connecting to the database right now.";
        QDamageTaken d = QDamageTaken.damageTaken;
        QPlayers p = QPlayers.players;
        List<Tuple> list = query
                .from(d)
                .leftJoin(p)
                .on(d.id.eq(p.id))
                .groupBy(p.name)
                .orderBy(d.amount.sum().desc())
                .limit(num)
                .list(p.name, d.amount.sum());

        return topListResponse("Damage Taken", list);
    }
}

File: src/main/kotlin/com/demonwav/statcraft/commands/sc/SCDamageTaken.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.commands.ResponseBuilderKt
import com.demonwav.statcraft.querydsl.QDamageTaken
import com.demonwav.statcraft.querydsl.QPlayers
import org.bukkit.command.CommandSender
import java.sql.Connection

class SCDamageTaken(plugin: StatCraft) : SCTemplate(plugin) {

    init {
        this.plugin.baseCommand.registerCommand("damagetaken", this)
    }

    override fun hasPermission(sender: CommandSender, args: Array<out String>?) = sender.hasPermission("statcraft.user.damagetaken")

    override fun playerStatResponse(name: String, args: List<String>, connection: Connection): String {
        try {
            val id = getId(name) ?: throw Exception()

            val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError
            val d = QDamageTaken.damageTaken

            val total = query.from(d).where(d.id.eq(id)).uniqueResult(d.amount.sum()) ?: 0

            return ResponseBuilderKt.build(plugin) {
                playerName { name }
                statName { "Damage Taken" }
                stats["Total"] = df.format(total)
            }
        } catch (e: Exception) {
            return ResponseBuilderKt.build(plugin) {
                playerName { name }
                statName { "Damage Taken" }
                stats["Total"] = 0.toString()
            }
        }

    }

    override fun serverStatListResponse(num: Long, args: List<String>, connection: Connection): String? {
        val query = plugin.databaseManager.getNewQuery(connection) ?: return databaseError
        val d = QDamageTaken.damageTaken
        val p = QPlayers.players

        val list = query
            .from(d)
            .leftJoin(p)
            .on(d.id.eq(p.id))
            .groupBy(p.name)
            .orderBy(d.amount.sum().desc())
            .limit(num)
            .list(p.name, d.amount.sum())

        return topListResponse("Damage Taken", list)
    }
}
